src/RadixDlt.NetworkGateway.Abstractions/TwoWayLinks/TwoWayLinkResolver.cs

```diff
@@ -0,0 +1,398 @@
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Newtonsoft.Json;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace RadixDlt.NetworkGateway.Abstractions.TwoWayLinks;
+
+public sealed class TwoWayLinkResolver
+{
+    private readonly IUnverifiedTwoWayLinksCollection _unverifiedTwoWayLinksCollection;
+    private readonly bool _resolveValidOnly;
+    private readonly bool _validateOnLedgerOnly;
+
+    public TwoWayLinkResolver(IUnverifiedTwoWayLinksCollection unverifiedTwoWayLinksCollection, bool resolveValidOnly, bool validateOnLedgerOnly)
+    {
+        _unverifiedTwoWayLinksCollection = unverifiedTwoWayLinksCollection;
+        _resolveValidOnly = resolveValidOnly;
+        _validateOnLedgerOnly = validateOnLedgerOnly;
+    }
+
+    public async ValueTask<Dictionary<EntityAddress, List<ResolvedTwoWayLink>>> Resolve(ICollection<EntityAddress> entityAddresses)
+    {
+        var results = new Dictionary<EntityAddress, List<ResolvedTwoWayLink>>();
+
+        foreach (var entityAddress in entityAddresses)
+        {
+            if (!_unverifiedTwoWayLinksCollection.TryGetTwoWayLinks(entityAddress, out var unverifiedTwoWayLinks))
+            {
+                continue;
+            }
+
+            foreach (var unverifiedTwoWayLink in unverifiedTwoWayLinks)
+            {
+                IEnumerable<ResolvedTwoWayLink> resolved = unverifiedTwoWayLink switch
+                {
+                    DappAccountTypeUnverifiedTwoWayLink dappAccountType => Resolve(entityAddress, dappAccountType),
+                    DappClaimedEntitiesUnverifiedTwoWayLink dappClaimedEntities => Resolve(entityAddress, dappClaimedEntities),
+                    DappClaimedWebsitesUnverifiedTwoWayLink dappClaimedWebsites => await Resolve(entityAddress, dappClaimedWebsites),
```

Comment:
We should separate / delay claimed website resolution into a separate ticket 👍

Reply:
Yeah, definitely. It was just a quick & dirty implementation to validate if it even works. As TODO says it's not meant to reach the production.